src/main/resources/templates/java/test/test/${packageNameFolder}/testutils/ClassNameScanner.java

```diff
@@ -0,0 +1,235 @@
+
+				if(hasWrongCase) {
+					scanResultType = classIsCorrectlyPlaced 
+							? ScanResultType.WRONGCASE_CORRECTPLACE
+							: ScanResultType.WRONGCASE_MISPLACED;
+					scanResultMessage = classIsCorrectlyPlaced
+							? "The exercise expects a class with the name " + expectedClassName
+									+ ". We found that you implemented a class " + observedClassName + ", which deviates from the expectation."
+									+ " Please check for wrong upper case / lower case lettering."
+							: "The exercise expects a class with the name " + expectedClassName + " in the package " + expectedPackageName
+									+ ". We found that you implemented a class " + observedClassName + ", in the package " + observedClassName
+									+ ", which deviates from the expectation."
+									+ " Please check for wrong upper case / lower case lettering and make sure you place it in the correct package.";
+					break;
+				} else if(hasTypos) {
+					scanResultType = classIsCorrectlyPlaced 
+							? ScanResultType.TYPOS_CORRECTPLACE
+							: ScanResultType.TYPOS_MISPLACED;
+					scanResultMessage = classIsCorrectlyPlaced
+							? "The exercise expects a class with the name " + expectedClassName
+									+ ". We found that you implemented a class " + observedClassName + ", which deviates from the expectation."
+									+ " Please check for typos in the class name."
+							: "The exercise expects a class with the name " + expectedClassName + "in the package " + expectedPackageName
+									+ ". We found that you implemented a class " + observedClassName + ", in the package " + observedPackageName
+									+ ", which deviates from the expectation."
+									+ " Please check for typos in the class name and make sure you place it in the correct package.";
+					break;
+				} else {
+					scanResultType = ScanResultType.NOTFOUND;
+					scanResultMessage = "You have implemented " + observedClassName + " in the package " + observedClassName
+							+ ". This class is not expected in the exercise."
+							+ "\n The assignment expects the following classes and in the according packages: \n";
+					
+					// Append the expected classes and package names to the result message
+					for(String expectedClassName : expectedClasses.keySet()) {
+						scanResultMessage += "Class: " + expectedClassName +
+											" Package:" + expectedClasses.get(expectedClassName) +",\n";
+					} 
+				}
+			}
+		}
+		
+		return new ScanResult(scanResultType, scanResultMessage);
+	}
+	
+	/**
+	 * This method retrieves the actual type names and their packages by walking the project file structure.
+	 * The root node (which is the assignment folder) is defined in the pom.xml file of the project.
+	 * @return The JSON object containing the type names as keys and the type packages as values.
+	 */
+	private HashMap<String, String> retrieveObservedClasses() {
```

Comment:
What happens if the class is present twice in the project (e.g. the student accidentally copied to another location) (especially if one of them is in the correct location)? I assume that an wrong copy might be returned in the HashMap if its checked after the correct one which leads to an error even if the class is also present at the expected location, is this correct?

Reply:
Good finding. I guess we need to include the package name together with the class name.
@kristiandimo please have a look

Reply:
Good point, will look into it!